check/src/unify_type.rs

```diff
@@ -163,6 +162,38 @@ impl<'a> Unifiable<State<'a>> for ArcType {
     }
 }
 
+fn try_remove_aliases(reduced_aliases: &mut Vec<Symbol>,
+                      env: &TypeEnv,
+                      typ: &ArcType)
+                      -> Result<Option<ArcType>, TypeError<Symbol>> {
+    if let Some(alias_id) = typ.alias_ident() {
+        if reduced_aliases.iter().any(|name| name == alias_id) {
+            return Err(TypeError::SelfRecursive(alias_id.clone()));
+        }
+        reduced_aliases.push(alias_id.clone());
+    }
+
+    match resolve::remove_alias(env, typ)? {
+        Some(mut typ) => {
+            loop {
+                if let Some(alias_id) = typ.alias_ident() {
+                    if reduced_aliases.iter().any(|name| name == alias_id) {
+                        return Err(TypeError::SelfRecursive(alias_id.clone()));
+                    }
+                    reduced_aliases.push(alias_id.clone());
+                }
+
+                match resolve::remove_alias(env, &typ)? {
+                    Some(new_typ) => typ = new_typ,
+                    None => break,
+                }
+            }
+            Ok(Some(typ))
+        }
+        None => Ok(None),
+    }
+}
```

Comment:
It can probably be simplified but the current complexity/duplication seems to be coming from the fact that `Ok (None)` is returned if no alias could be expanded and `Ok (Some (typ))` is returned if at least one alias was expanded